proposals/0012-introduce-reactNativeMetadata.md

```diff
@@ -0,0 +1,447 @@
+
+### New Architecture support
+
+One of the primary driver of this proposal is the [New React Native Architecture support](https://reactnative.dev/docs/the-new-architecture/landing-page). Specifically we believe that the ecosystem is currently affected by:
+* Lack of a declarative way to define if a library is compatible with the New Architecture
+  * As described in the [rationale](#rationale), tooling needs to infer this information from the presence of the `codegenConfig` field in the `package.json` of a library. Libraries might decide not to use Codegen and still be compatible with the New Architecture, so this method is not exhaustive.
+  * The alternative at this stage is to inspect the code and check whether the API callsites are New Architecture or Old Architecture compatible.
+* Lack of a declarative way to enable the New Architecture on both platforms
+  * Currently the New Architecture is enabled by a Gradle Property `newArchitectureEnabled` on Android in the `gradle.properties` file and by invoking `RCT_NEW_ARCH_ENABLED=1 bundle exec pod install` on iOS.
+  * As of today, there is no way to enable the New Architecture for an app project for both platforms
+  * Moreover, the difference in how the New Architecture is enabled leads to a scenario where you can't statically know if an app supports New Architecture or not (as the New Architecture support is known at  `pod install` time and is not codified in the codebase)
+
+Therefore we propose to add the `newArchitecture` section to the `reactNativeManifest.capabilities` of **both apps and libraries** with the following semantic:
+* For Apps: `newArchitecture.enabled==true` means that the app wants to use the New Architecture.
+* For Libraries: `newArchitecture.enabled==true` means that the library is compatible with the New Architecture.
+
+Tools can be built on top of this information to check that an app with `newArchitecture.enabled==true` is not accepting libraries with `newArchitecture.enabled==false` and to warn against library that don't have the key specified.
+
+The setup would look as follows for both apps and libraries:
+
+```json
+{
+  "reactNativeManifest": {
+    "capabilities": {
+      "newArchitecture": {
+        "enabled": true
+      }
+    }
+  }
+}
+```
+
+This section will allow also for split configuration between platforms:
+
+```json
+{
+  "reactNativeManifest": {
+    "android": {
+      "capabilities": {
+        "newArchitecture": {
+          "enabled": true
+        }
+      }
+    },
+    "ios": {
+      "capabilities": {
+        "newArchitecture": {
+          "enabled": true
+        }
+      }
+    }
+  }
+}
+```
+
+If omitted, the `newArchitecture.enabled` would have a value of `false` (this may change in the future).
+
+The precedence rules are that platform-specific settings override general settings. So, for example:
+
+```json
+{
+  "reactNativeManifest": {
+    "capabilities": {
+      "newArchitecture": {
```

Comment:
Thanks @acoates-ms! I guess you mean for libraries, right? 

That's a good point. What about changing `newArchitecture` in `architecture` which is an array of `["old", "new"]`?
In version 1.0.0 of this ReactNative Manifest, if not specified, it assumes `["old"]`, so only paper/Native Modules.

For apps, the array must have a single entry and it should be either `new` or `old` and it determines how to build the app.

If it makes sense, I'll update the proposal.

Reply:
If it's important for a library to signal that it (also) supports the old architecture, wouldn't it be better modelled as another "capability"? I.e.

```json
{
  "reactNativeManifest": {
    "capabilities": {
      "newArchitecture": {
        "enabled": true
      },
      "oldArchitecture": {
        "enabled": true
      }
    }
  }
}
```

Reply:
@kraenhansen it could. I think that it depends whether the think that the architecture capabilities might be further extended. Probably, you are right that the separate approach make it more extensible in the future.

Reply:
Applied in [ccba7f4](https://github.com/react-native-community/discussions-and-proposals/pull/717/commits/ccba7f4bb6f4650b30a0e61993851b3e0fd2e4e9)